docs/blog/2021-06-16-tailor-your-slack-audit-sink-messages.md

```diff
@@ -0,0 +1,149 @@
+    typed_config:
+      “@type”: types.google.com/clutch.config.service.auditsink.slack.v1.SlackConfig
+      ...
+      filter:
+        rules:
+          - field: SERVICE
+          text: clutch.k8s.v1.K8sAPI
+        overrides:
+          - full_method: /clutch.k8s.v1.K8sAPI/ResizeHPA
+          - message: |
+            *Min size*: [[.Request.sizing.min]]
+            *Max size*: [[.Request.sizing.max]]
+```
+
+A few things to highlight from the example:
+  * `.Request.<field name>` is used to retrieve data from the API request.
+  * Clutch-specific [templating token](https://github.com/lyft/clutch/blob/0aa1c00b37513900c351be1106cc131498b1aad0/backend/gateway/config.go#L110-L122) is used in lieu of the standard Golang template [Actions](https://golang.org/pkg/text/template/#hdr-Actions) syntax. This is to avoid conflicts during preprocessing of the clutch-config.
+  * Slack [markdown](https://api.slack.com/reference/surfaces/formatting#basics) formatting can be added for useful visual highlights to the custom message.
+  * Users can provide one custom message per gRPC method.
+
+That minimal configuration is all that’s needed to set up this feature for any Clutch workflow!
+
+<img alt="Clutch Custom Slack Audit Message" src="https://user-images.githubusercontent.com/39421794/122236354-734b7380-ce8c-11eb-97e4-146222eda962.png" />
+
+## Under the Hood
+
+In the Slack sink, we check if a user has provided a custom Slack message for a given workflow  event and if so, inject the metadata values from the audit event into the custom message and [execute](https://golang.org/pkg/text/template/#Template.Execute) the message template.
+
+As mentioned earlier, one of our goals for the customized Slack message feature was to reduce code duplication. By saving the API request/response body as Any type in the audit event, we can retrieve the metadata values for any custom message with just this flow [here](https://github.com/lyft/clutch/blob/9765a88f042c306b0a9d63fb55a8b2ed552b1c18/backend/service/auditsink/slack/slack.go#L174-L202) in the Slack sink. Let’s focus on the important pieces.
+
+```go
+reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+```
+
+The aim here is to unpack the metadata (Any type) into its underlying message. We looked into using anypb’s [UnmarshalNew](https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb#Any.UnmarshalNew) but ultimately decided upon protojson’s [Marshal](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#Marshal). The main reason for this choice was that we wanted to use the json format as the basis for the context in order to simplify the metadata structure a bit. For example, protojson flattens oneOf fields into its top level field. Since custom messages need to specify what exact data is needed from the request/response, simplifications like these can help make writing a custom message simpler/easier.
```

Comment:
```suggestion
The aim here is to unpack the metadata (Any type) into its underlying message. We looked into using anypb’s [UnmarshalNew](https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb#Any.UnmarshalNew) but ultimately decided upon protojson’s [Marshal](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#Marshal). The main reason for this choice was that we wanted to use the json format as the basis for the context in order to simplify the metadata structure a bit. For example, protojson flattens oneOf fields into its top level field. Since custom messages need to specify what exact data is needed from the request/response, simplifications like these can help make writing a custom message easier.
```